File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/RecoveryUpload.kt
// Kotlin mirror of the UploadRecoveryRecord canonical in
// packages/protocol/src/auth.ts (canonicalUploadRecoveryRecord) +
// packages/control-plane/src/webauthnRecovery.ts handleUploadWebauthnRecovery.
//
// Cloud recovery upload is IRK-SIGNED. The wire body is
//   { request: { username, credentialId, wrappedUmk, issuedAt,
//                 wrappedAcmeAccountKey? }, signature }
// and the Worker base64-decodes `wrappedUmk` → bytes, computes
//   wrappedUmkHashHex = sha256hex(bytes)
// then verifies `signature` (hex, ed25519 by the account IRK) over the
// canonical bytes below. The signature commits to the HASH of the
// ciphertext (not the ciphertext itself) so the canonical stays small,
// while .com still re-derives the hash from the bytes on the wire and
// rejects any attacker-substituted ciphertext.
//
// The canonical `|`-joined field order MUST match the Worker byte-for-byte
// or server verification fails. `wrappedAcmeAccountKey` is ciphertext only
// and is NOT part of the canonical (tampering breaks account-key recovery
// but can never forge it).

package com.flagshipserver.app.core

import com.google.crypto.tink.subtle.Ed25519Sign
import java.security.MessageDigest

object RecoveryUpload {
    const val CANONICAL_TAG = "flagship/upload-recovery-record/v1"

    /** Canonical bytes the account IRK signs over. Mirrors
     *  canonicalUploadRecoveryRecord in @flagship/protocol:
     *  `tag|username|credentialId|wrappedUmkHashHex|issuedAt`. */
    fun canonicalBytes(
        username: String,
        credentialId: String,
        wrappedUmkHashHex: String,
        issuedAt: Long,
    ): ByteArray = listOf(
        CANONICAL_TAG,
        username,
        credentialId,
        wrappedUmkHashHex,
        issuedAt.toString(),
    ).joinToString("|").toByteArray(Charsets.UTF_8)

    /** Sign with the account's CURRENT IRK — the only key whose pubkey the
     *  Worker looks up (from the usernames row) to verify the upload. */
    fun sign(
        irk: Ed25519Sign,
        username: String,
        credentialId: String,
        wrappedUmkHashHex: String,
        issuedAt: Long,
    ): ByteArray = irk.sign(
        canonicalBytes(username, credentialId, wrappedUmkHashHex, issuedAt),
    )

    /** Lowercase sha256-hex of the raw wrapped-UMK ciphertext bytes — the
     *  exact value the Worker recomputes from `wrappedUmk` (after base64
     *  decode) and checks the signature against. */
    fun wrappedUmkHashHex(wrappedUmkBytes: ByteArray): String =
        HexUtil.encode(MessageDigest.getInstance("SHA-256").digest(wrappedUmkBytes))
}

File: apps/mobile/ios/Sources/Flagship/RecoveryUpload.swift
import Foundation
import CryptoKit
import FlagshipCore

/// Swift mirror of @flagship/protocol's `UploadRecoveryRecord` canonical
/// bytes + signer. Keep in sync with `packages/protocol/src/auth.ts`
/// (`canonicalUploadRecoveryRecord` / `signUploadRecoveryRecord`).
///
/// The phone uploads a cloud-recovery shard by POSTing a SIGNED envelope
/// to `POST /api/recovery`. The Worker base64-decodes `wrappedUmk`,
/// computes `wrappedUmkHashHex = sha256hex(those bytes)`, and verifies the
/// signature over the canonical string below under the account IRK. The
/// ciphertext is hashed (not signed directly) so the canonical bytes stay
/// small, and so .com cannot be tricked into storing attacker-substituted
/// ciphertext under a victim's passkey.
///
/// The `|`-joined field order MUST match the Worker byte-for-byte or
/// `verifyUploadRecoveryRecord` on the server fails.
public enum RecoveryUpload {
    /// `flagship/upload-recovery-record/v1`, same tag the Worker uses.
    public static let canonicalTag = "flagship/upload-recovery-record/v1"

    /// `flagship/upload-recovery-record/v1|<username>|<credentialIdHex>|<wrappedUmkHashHex>|<issuedAt>`.
    /// `credentialIdHex` and `wrappedUmkHashHex` are kept verbatim (the
    /// Worker re-derives the hash from the wire ciphertext and compares the
    /// signed value against it; the comparison is case-sensitive on the
    /// server side, so callers pass the exact hex they signed).
    public static func canonical(
        username: String,
        credentialIdHex: String,
        wrappedUmkHashHex: String,
        issuedAt: Int64
    ) -> Data {
        Data(
            [
                canonicalTag,
                username,
                credentialIdHex,
                wrappedUmkHashHex,
                String(issuedAt),
            ].joined(separator: "|").utf8
        )
    }

    /// Sign the canonical bytes with the account's CURRENT IRK and return
    /// the Ed25519 signature as lowercase hex (the wire encoding the Worker
    /// expects in the envelope's `signature` field).
    public static func sign(
        username: String,
        credentialIdHex: String,
        wrappedUmkHashHex: String,
        issuedAt: Int64,
        irk: Curve25519.Signing.PrivateKey
    ) throws -> String {
        let sig = try irk.signature(
            for: canonical(
                username: username,
                credentialIdHex: credentialIdHex,
                wrappedUmkHashHex: wrappedUmkHashHex,
                issuedAt: issuedAt
            )
        )
        return HexUtil.encode(sig)
    }

    /// SHA-256 of the wrapped-UMK ciphertext bytes, lowercase hex. The
    /// caller base64-decodes the single self-contained `wrappedUmk` blob
    /// (nonce‖ct‖tag) and feeds the raw bytes here so the hash matches what
    /// the Worker computes from the same wire value.
    public static func wrappedUmkHashHex(_ wrappedUmkBytes: Data) -> String {
        let digest = SHA256.hash(data: wrappedUmkBytes)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
